src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs

```diff
@@ -791,18 +796,41 @@ protected TNode AddError<TNode>(TNode nodeOrToken, ErrorCode code, params object
             // token, and we will want to place the diagnostic on it.  Otherwise, we want to place it on the true 'next
             // token' the parser is currently pointing at.
 
-            if (!nodeOrToken.ContainsSkippedText)
+            if (!missingNodeOrToken.ContainsSkippedText)
             {
                 // Simple case this node/token does not contain any skipped text.  Place the diagnostic at the start of
                 // the token that follows.
-                var (offset, width) = this.GetDiagnosticSpanForMissingToken();
-                return WithAdditionalDiagnostics(nodeOrToken, MakeError(nodeOrToken.FullWidth + offset, width, code, args));
+                return getOffsetAndWidthBasedOnPriorAndNextTokens();
             }
             else
             {
                 // Complex case.  This node or token contains skipped text.  Place the diagnostic on the skipped text.
-                var (offset, width) = getOffsetAndWidthOfSkippedToken();
-                return WithAdditionalDiagnostics(nodeOrToken, MakeError(offset, width, code, args));
+                return getOffsetAndWidthOfSkippedToken();
+            }
+
+            (int offset, int width) getOffsetAndWidthBasedOnPriorAndNextTokens()
+            {
+                // If the previous token has a trailing EndOfLineTrivia, the missing token diagnostic position is moved
+                // to the end of line containing the previous token and its width is set to zero. Otherwise we squiggle
+                // the token following the missing token (the token we're currently pointing at).
+
+                var trivia = _prevTokenTrailingTrivia;
+                var triviaList = new SyntaxList<CSharpSyntaxNode>(_prevTokenTrailingTrivia);
```

Comment:
nit: this could use `trivia` instead of `_prevTokenTrailingTrivia`. However, this is the same in the original code, so no need to make a change.